Feature: Implement three kernels in C++ CUDA

drewsilcock opened the pull request:
Implement:

- Continuity
- Boundary conditions
- Next
- Getting values out of CUDA

This is not finalised yet - it's not producing a correct checksum currently which I'm still debugging.

drewsilcock committed:
Implement boundary conditions kernel in CUDA C++.

Slightly more complicated due to the need for double buffering on ua and va due
to possible data races.

drewsilcock committed:
Add missing delete for double buffer ua and va.

drewsilcock committed:
Implement next kernel in CUDA C++.

drewsilcock committed:
Implement continuity kernel in CUDA C++.

drewsilcock committed:
Add functions to retrieve data from CUDA device.

Data is retrieved from the CUDA device into a Fortran array via C
bindings.

drewsilcock committed:
Fix length of ierr array in allocation subroutine.

drewsilcock committed:
Write out CUDA var checksums and output CUDA dat files.

drewsilcock committed:
Launch kernels in 2D as per Pete's suggestion.

drewsilcock committed:
Use CUDACHECK for all CUDA calls.

drewsilcock committed:
Move CUDACHECK into separate header.

This way it can also be used by the fortran 2D array class.

drewsilcock committed:
Use device constant cache and cudaMemcpyToSymbol for constants.

drewsilcock committed:
Ignore Vim files from version control.

drewsilcock committed:
Format CUDA C++ file with clang-format.

ptheywood commented on cuda/cuda_nemolite2d.cu:
```diff
@@ -4,12 +4,38 @@
+
+  // Suggest block dimensions. Threads per block must not exceed 1024 on most
+  // hardware, registers will probably be a limiting factor.
+  dim3 blocksize(2, 2);
```
2x2 is too small for good performance ( atleast 32 threads will be launched). 

Additionally, different kernels use differnt numbers of registers, so the "best" blocksize may be different per kernel. This can either be found through trial and error / looking at profiler output and optimising for occupancy. The occupancy calculator api is an option, though it's unclear how it works for 2D launch. I believe you tell it the total number fo threads you want and it returns the number of threads per block (which you are able to divide into 2D as you wish)

drewsilcock committed:
Remove __device__ from CUDA device constants.

drewsilcock committed:
Fix kernel launching and thread (x,y) calculation.

All values passed into kernel launch should be by value, not by reference or
using pointer due to separate address spaces.

Use 16,16 instead of 2,2 as starting suggestion for kernel dimensions.

drewsilcock committed:
Fix calloc in Fortran array wrapper constructor.

Also add debug printing to various methods.

drewsilcock committed:
Fix getting data out of FortranArray2D.

Destructor was being called prematurely. You now need to manually call a method
to free the data (for which there is a macro to free + delete).

drewsilcock commented:
This is now running and producing legitimate output. The checksums are still 0 but I think this is just because the momentum kernel isn't implemented which is what calculates ua and va.

ptheywood committed:
Swich to cuda occupancy api calls for blockdims

drewsilcock committed:
Merge pull request #30 from ptheywood/feature/implement-three-kernels-in-cpp-cuda

Swich to cuda occupancy api calls for blockdims
